simulation/conveyor/model.py

```diff
@@ -0,0 +1,325 @@
+
+    def _stateTransfer(self, action: str) -> None:
+        try:
+            self._state = _model_automata[self._state][action]
+        except KeyError:
+            raise AutomataException(
+                f'{self._model_id}: Invalid action `{action}` in state `{self._state}`;\n  speed - {self._speed}m/s\n  cps - {self._checkpoints};\n  objs - {self._object_positions}')
+
+
+    def nearestObject(self, pos: float, after=None, speed=None, not_exact=False,
+                      preference='nearest') -> Optional[Tuple[Any, float]]:
+        if len(self._object_positions) == 0:
+            return None
+
+        if after is not None:
+            if speed is None:
+                speed = self._speed
+            objs = shift(self._object_positions, after * speed)
+        else:
+            objs = self._object_positions
+
+        res = search_pos(objs, pos, preference=preference)
+        if res is not None:
+            (oid, o_pos), idx = res
+            if not_exact and o_pos == pos:
+                if preference == 'prev':
+                    idx -= 1
+                elif preference == 'next':
+                    idx += 1
+                else:
+                    raise Exception('please dont use nearest with not exact')
+                if idx < 0 or idx > len(objs):
+                    return None
+                oid, o_pos = objs[idx]
+            return self._objects[oid], o_pos
+        return None
+
+    def putObject(self, obj_id: int, obj: Any, pos: float, soft_collide=True, return_nearest=False):
+        assert obj_id not in self._objects, "Clashing object ID!"
+        pos = round(pos, POS_ROUND_DIGITS)
+
+        nearest = None
+        if len(self._objects) > 0:
+            (n_obj_id, n_pos), n_idx = search_pos(self._object_positions, pos)
+            if n_pos == pos:
+                if soft_collide:
+                    print(f'{self._model_id}: TRUE COLLISION: #{obj_id} and #{n_obj_id} on {pos}')
+                    i = n_idx
+                    p_pos = pos
+                    while i < len(self._object_positions) and self._object_positions[i][1] >= p_pos:
+                        p_pos = round(p_pos + SOFT_COLLIDE_SHIFT, POS_ROUND_DIGITS)
+                        self._object_positions[i] = (self._object_positions[i][0], p_pos)
+                        i += 1
+                else:
+                    raise CollisionException((obj, self._objects[n_obj_id], pos, self._model_id))
+            elif n_pos < pos:
+                n_idx += 1
+            nearest = (n_obj_id, n_pos)
+        else:
+            n_idx = 0
+
+        self._objects[obj_id] = obj
+        self._object_positions.insert(n_idx, (obj_id, pos))
+        self._stateTransfer('change')
+
+        if return_nearest:
+            return nearest
+
+    def removeObject(self, obj_id: int):
+        pos_idx = None
+        for (i, (oid, pos)) in enumerate(self._object_positions):
+            if oid == obj_id:
```

Comment:
I think something like [this](https://www.programiz.com/python-programming/methods/list/index) will be more readable. You may just drop it if you are not agree